SWIPs/swip-22-multiple_payment_processing_support.md

```diff
@@ -0,0 +1,464 @@
+
+Experimentation is needed to determine the best time for nodes to engage in payment module negotiation.
+
+### Multiple payment modules support
+When allowing multiple payment modules, it is essential for nodes to communicate their preference. There are three dimensions to take into consideration for the payment preference: 
+
+1. Currency to use
+2. Price oracle to use
+3. Payment provider to use
+
+We propose a standardized way to communicate these preferences and an algorithm to resolve any two preferences lists.
+
+For any dimension, nodes list the relative importance of them. Consequently, nodes specify their accepted options within each dimension. Importance is communicated through Swarm by any positive integer.
+
+The following is an example of the proposed configuration:
+
+```yaml
+# The relative importance of each dimension for the node
+dimensions:
+    currency: 70
+    provider: 20
+    oracle: 10
+
+# Supported payment options by the node.
+# A payment module is defined by a currency, the supported providers and oracles. Weights are the values associated to each provider.
+# or oracle and it indicates the preference for that option.
+currencies:
+    rif:
+        weight: 36
+        providers:
+            lumino: 15
+            raiden: 45
+        oracles:
+            rifOracleA: 10
+            rifOracleB: 70
+            rifOracleC: 60
+    xdai:
+        weight: 9
+        providers:
+            lumino: 10
+        oracles:
+            xDaiOracle: 20
+    dai:
+        weight: 5
+        providers:
+            raiden: 80
+        oracles:
+            daiOracleA: 30
+            daiOracleB: 20
+    eth:
+        weight: 40
+        providers:
+            swap: 90
+        oracles:
+            ethOracleA: 90
+            ethOracleB: 10
+```
+
+When two nodes (A and B) establish a connection they independently execute the following payment method selection algorithm, using as input the configuration defined previoulsy:
+
+1. Based on the node configuration generate a set of triplets in the form ```[currency, provider, oracle]```. As an example let's take from the previous configuration the ```rif``` entry. From that entry we can build the triplets `[rif, lumino, rifOracleA]`, `[rif, lumino, rifOracleB]`, `[rif, lumino, rifOracleC]`,`[rif, raiden, rifOracleA]`, `[rif, raiden, rifOracleB]`, `[rif, raiden, rifOracleC]`, but triplets are generated for the other currencies as well. 
```

Comment:
What are the requirements for nodes communicating across different implementations of Swarm? E.g. wire format, protobuf etc... is that out of scope of this document?

Reply:
From my point of view that is out of the scope of this document.